continuous_probabilistic_methods.py or continuous_probabilistic_methods.ipynb 

0. Define a function named get_lower_and_upper_bounds that has two arguments. The first argument is a pandas Series. The second argument is the multiplier, which should have a default argument of 1.5.


1. Using lemonade.csv dataset and focusing on continuous variables:
    - Use the IQR Range Rule and the upper and lower bounds to identify the lower outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense?Which outliers should be kept?
    - Use the IQR Range Rule and the upper and lower bounds to identify the upper outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense?Which outliers should be kept?
    - Using the multiplier of 3, IQR Range Rule, and the lower and upper bounds, identify the outliers below the lower bound in each colum of lemonade.csv. Do these lower outliers make sense?Which outliers should be kept?
    - Using the multiplier of 3, IQR Range Rule, and the lower and upper bounds, identify the outliers above the upper_bound in each colum of lemonade.csv. Do these upper outliers make sense? Which outliers should be kept?


2. Identify if any columns in lemonade.csv are normally distributed. For normally distributed columns:
    - Use a 2 sigma decision rule to isolate the outliers.
    - Do these make sense?
    - Should certain outliers be kept or removed?


3. Now use a 3 sigma decision rule to isolate the outliers in the normally distributed columsn from lemonade.csv

In [1]:
import pandas as pd
import numpy as np

#### Define a function named get_lower_and_upper_bounds that has two arguments. The first argument is a pandas Series. The second argument is the multiplier, which should have a default argument of 1.5.

In [13]:
def get_lower_and_upper_bounds(series, multiplier=1.5):
    q1 = np.quantile(series, .25)
    q3 = np.quantile(series, .75)
    iqr = q3 - q1 
    upper_bound = q3 + (multiplier * iqr)
    lower_bound = q1 - (multiplier * iqr)
    return upper_bound, lower_bound

#### Using lemonade.csv dataset and focusing on continuous variables:

In [8]:
df = pd.read_csv("lemonade.csv")

In [9]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 365 entries, 0 to 364
Data columns (total 7 columns):
Date           365 non-null object
Day            365 non-null object
Temperature    365 non-null float64
Rainfall       365 non-null float64
Flyers         365 non-null int64
Price          365 non-null float64
Sales          365 non-null int64
dtypes: float64(3), int64(2), object(2)
memory usage: 20.1+ KB


Date        Day  Temperature  Rainfall  Flyers  Price  Sales
0  1/1/17     Sunday         27.0      2.00      15    0.5     10
1  1/2/17     Monday         28.9      1.33      15    0.5     13
2  1/3/17    Tuesday         34.5      1.33      27    0.5     15
3  1/4/17  Wednesday         44.1      1.05      28    0.5     17
4  1/5/17   Thursday         42.4      1.00      33    0.5     18

> Why is this dataset called lemonade? 

#### Use the IQR Range Rule and the upper and lower bounds to identify the lower outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense?Which outliers should be kept?
    

In [39]:
print("Lower bounds:\n")
for col in df.select_dtypes(np.number):
    print("The lower bound for {} is {}".format(col, get_lower_and_upper_bounds(df[col])[1]))
    print("The outliers are as follows:")
    for value in df[col]:
        if value <= get_lower_and_upper_bounds(df[col])[1]:
            print(value)
    print()

Lower bounds:

The lower bound for Temperature is 16.700000000000003
The outliers are as follows:
15.1

The lower bound for Rainfall is 0.26
The outliers are as follows:

The lower bound for Flyers is 4.0
The outliers are as follows:
-38

The lower bound for Price is 0.5
The outliers are as follows:
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5

> Temperature: This looksl like a reasonable fluctuation and should be kept

> Rainfall: No lower bound outliers

> Flyers: The outlier is a negative value. Should be removed
    
> Price: The outlier of 0.5 is many and therefore, should be kept
    
> Sales: No lower bound outliers    

#### Use the IQR Range Rule and the upper and lower bounds to identify the upper outliers of each column of lemonade.csv, using the multiplier of 1.5. Do these lower outliers make sense?Which outliers should be kept?

In [44]:
print("Upper bounds:\n")
for col in df.select_dtypes(np.number):
    print("The upper bound for {} is {}".format(col, get_lower_and_upper_bounds(df[col])[0]))
    print("The outliers are as follows:")
    for value in df[col]:
        if value >= get_lower_and_upper_bounds(df[col])[0]:
            print(value)
    print()

Upper bounds:

The upper bound for Temperature is 104.7
The outliers are as follows:
212.0

The upper bound for Rainfall is 1.3
The outliers are as follows:
2.0
1.33
1.33
1.54
1.54
1.54
1.33
1.33
1.67
1.43
1.43
1.54
1.33
1.33
1.54
1.82
1.43
1.82
1.33
1.43
1.54
1.33
1.43
1.33
1.54
1.43
1.43
2.5

The upper bound for Flyers is 76.0
The outliers are as follows:
77
76
80
76

The upper bound for Price is 0.5
The outliers are as follows:
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0.5
0

> Temperature: A temperature of 212 seems unreasonable. Remove
> Rainfall: 

#### Using the multiplier of 3, IQR Range Rule, and the lower and upper bounds, identify the outliers below the lower bound in each colum of lemonade.csv. Do these lower outliers make sense?Which outliers should be kept?

#### Using the multiplier of 3, IQR Range Rule, and the lower and upper bounds, identify the outliers above the upper_bound in each colum of lemonade.csv. Do these upper outliers make sense? Which outliers should be kept?